### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by sergej-tschak**
- throw errors with unified error object type
- handle network errors, token refresh error, wrong data sent errors, connection errors
- add timeout to axios requests

### CODE REVIEW

File: src/errors/index.ts [hunk 1]
```diff
@@ -1,10 +1,16 @@
...
+  ResponseFailed = 'Response Failed',
 }
 
 export type TimeTacApiError = {
   reason: ErrorReason;
   response?: unknown;
   _plainError?: string;
 };
+
+export type TTError = {
```
Comment (tomi-timetac) [inline src/errors/index.ts:12]: Any reason its named TTError? As this is not really descriptive.

File: src/index.ts [hunk 2]
```diff
@@ -134,6 +135,7 @@ export default class Api {
   public setBaseUrl() {
     setAxiosDefaults({
       baseURL: `${this.config.settings.https == true ? 'https' : 'http'}://${this.config.settings.host ?? DEFAULT_HOST}`,
+      timeout: this.config.settings.timeout,
```
Comment (sergej-tschak) [inline src/index.ts:138]: recheck pls if this one works

File: src/utils/axiosSetup.ts [hunk 3]
```diff
@@ -1,65 +1,91 @@
...
-type interceptorParams = {
+export type InterceptorParams = {
   state: ApiState;
   config: ConfigProvider;
   authentication: AuthenticationEndpoint;
 };
 
+const requestInterceptor = (config: AxiosRequestConfig) => {
+  // axios.defaults do not automatically apply here, so set timeout manually
+  config.timeout = axios.defaults.timeout;
```
Comment (sergej-tschak) [inline src/utils/axiosSetup.ts:17]: the same with timeout here

**Comment by sergej-tschak** [comment]
@benj56 could you please check this things:
- if timeout works
- if we have all needed use cases covered with tests

File: src/errors/index.ts [hunk 4]
```diff
@@ -1,10 +1,5 @@
...
-  Unkown = 'Unknow Error',
-  ReponseFailed = 'Reponse Failed',
-}
-
-export type TimeTacApiError = {
-  reason: ErrorReason;
-  response?: unknown;
-  _plainError?: string;
+export type ErrorFormat = {
+  statusCode: number | undefined;
```
Comment (AlGvrl) [inline src/errors/index.ts:2]:
```suggestion
  statusCode?: number;
```
Or does this not work?

File: src/errors/index.ts [hunk 5]
```diff
@@ -1,10 +1,5 @@
...
+  message: string | undefined;
```
Comment (AlGvrl) [inline src/errors/index.ts:3]:
```suggestion
  message?: string;
```
Same

File: src/index.ts (hunk 2 above)
Comment (AlGvrl) [reply src/index.ts:138]: +1 here, have some doubts about it possibly being `undefined`.

File: src/utils/axiosSetup.ts [hunk 6]
```diff
@@ -1,65 +1,91 @@
...
-      if (apiInstanceData.config.settings.autoRefreshToken && error.response) {
-        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
-        const untouchedRequest = error.config;
-        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-call,@typescript-eslint/strict-boolean-expressions
-        if (error.response.status === 497 && untouchedRequest.url.includes('auth/oauth2/token')) {
-          if (apiInstanceData.config.settings.onTokenRefreshedFailed) {
-            apiInstanceData.config.settings.onTokenRefreshedFailed();
+export const responseRejectedInterceptor = (interceptorParams: InterceptorParams) => async (error: AxiosError) => {
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions,@typescript-eslint/no-unsafe-member-access
+  if (interceptorParams.config.settings.autoRefreshToken && error.response) {
```
Comment (AlGvrl) [inline src/utils/axiosSetup.ts:26]: Can we maybe rename this to `shouldAutoRefreshToken` or something similar? Doesn't look like a bool to me.

File: src/utils/axiosSetup.ts [hunk 7]
```diff
@@ -1,65 +1,91 @@
...
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-member-access
+    const untouchedRequest = error.config as AxiosRequestConfig & { _retry: boolean };
```
Comment (AlGvrl) [inline src/utils/axiosSetup.ts:28]: Maybe the same "doesn't look bool enough" here for `_retry`, but I'm not sure if it's coming from our code or from the backend.

File: src/utils/axiosSetup.ts [hunk 8]
```diff
@@ -1,65 +1,91 @@
...
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/strict-boolean-expressions
+    if (error.response.status === 401 && !untouchedRequest._retry && !error.response.config.url?.includes('oauth2')) {
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+      untouchedRequest._retry = true;
+
+      // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+      if (!interceptorParams.state.refreshingToken) {
```
Comment (AlGvrl) [inline src/utils/axiosSetup.ts:36]: And maybe here (`isRefreshingToken`). Though here it's not as upsetting to me -- I didn't manage to misinterpret this one :smile:

File: src/utils/axiosSetup.ts [hunk 9]
```diff
@@ -1,65 +1,91 @@
...
+    message: error.message,
+    raw: error,
+  };
+
+  throw toThrow;
+};
+
+export const interceptor = (interceptorParams: InterceptorParams): void => {
+  axios.interceptors.request.use(requestInterceptor);
+  axios.interceptors.response.use(responseFulfilledInterceptor, responseRejectedInterceptor(interceptorParams));
```
Comment (AlGvrl) [inline src/utils/axiosSetup.ts:88]: One of those is not an interceptor. One of those takes a param and returns an interceptor.

File: src/utils/axiosSetup.ts [hunk 10]
```diff
@@ -1,65 +1,91 @@
...
+        return axios(untouchedRequest);
       }
-      // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access,@typescript-eslint/strict-boolean-expressions
-      throw error.response || error;
     }
-  );
+  }
+
+  const toThrow: ErrorFormat = {
+    statusCode: error.code != null && error.code.length > 0 ? parseInt(error.code) : undefined,
```
Comment (AlGvrl) [inline src/utils/axiosSetup.ts:78]: `parseInt("123onetwothree")` returns `123`. Is this what we want here, or would we rather treat this as invalid?

**Review by AlGvrl — APPROVED** [review]
A bunch of naming/code style comments; overall lgtm

File: src/utils/axiosSetup.ts (hunk 8 above)
Comment (benj56) [reply src/utils/axiosSetup.ts:36]: Unfortunately it's `refreshingToken: false | Promise<AxiosResponse<TokenResponse>>;`, that part was just moved from the old interceptor

File: src/index.ts [hunk 11]
```diff
@@ -137,6 +139,13 @@ export default class Api {
     });
   }
 
+  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+  public setTimeout(timeout: number) {
```
Comment (benj56) [inline src/index.ts:143]:
```suggestion
  public setTimeout(timeout: number): void {
```

**Review by benj56 — APPROVED** [review]

**Review by benj56 — DISMISSED** [review]

**Review by tomi-timetac — DISMISSED** [review]

**Review by tomi-timetac — DISMISSED** [review]

**Review by benj56 — APPROVED** [review]

**Review by benj56 — APPROVED** [review]

**Pull request merged**